---
title: "Airway RNA-Seq Data Anaysis Workflow"
author: "Chiranjit Mukherjee"
date: "11/29/2020"
output:
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
<br>

### Notes
This project uses the 'airway' package that summarizes an RNA-seq experiment wherein airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014). Glucocorticoids are used, for example, by people with asthma to reduce inflammation of the airways. In the experiment, four primary human airway smooth muscle cell lines were treated with 1 micromolar dexamethasone for 18 hours. For each of the four cell lines, we have a treated and an untreated sample.

<br>

#### Load libraries
```{r, message=FALSE}
library(tidyverse)
library(airway)
library(DESeq2)
library(ggplot2)
library(apeglm)
library(AnnotationHub)
library(ReportingTools)
```

<br>

#### RStudio Session Information
```{r}
sessionInfo()
```



#### Load the airway data
```{r}
data("airway")
airway
```
The RangedSummarizedExperiment class is a matrix-like container where rows represent ranges of interest (as a GRanges or GRangesList object) and columns represent samples (with sample data summarized as a DataFrame). A RangedSummarizedExperiment contains one or more assays, each represented by a matrix-like object of numeric or other mode.
RangedSummarizedExperiment is a subclass of SummarizedExperiment and, as such, all the methods documented in ?SummarizedExperiment also work on a RangedSummarizedExperiment object. The methods documented below are additional methods that are specific to RangedSummarizedExperiment objects.


<br>

#### Specify that 'untrt' is the reference level for the dex variable
```{r}
airway$dex <- relevel(airway$dex, "untrt")
levels(airway$dex)
```

<br>

```{r}
head(assay(airway))
```


<br>

#### Check the millions of fragments that uniquely aligned to the genes
```{r}
round(colSums(assay(airway))/1e6, 1)
```

<br>

#### Inspect the information about the samples
```{r}
table(airway$dex)
colData(airway)
```

<br>

#### Prepare data for differential expression analysis using DESeq2
```{r}
dds <- DESeqDataSet(airway, design = ~ cell + dex) # controlling for cell lines while testing for differences due to treatment
```
DESeqDataSet is a subclass of RangedSummarizedExperiment, used to store the input values, intermediate calculations and results of an analysis of differential expression. The DESeqDataSet class enforces non-negative integer values in the "counts" matrix stored as the first element in the assay list. In addition, a formula which specifies the design of the experiment must be provided. The constructor functions create a DESeqDataSet object from various types of input: a RangedSummarizedExperiment, a matrix, count files generated by the python package HTSeq, or a list from the tximport function in the tximport package. 

<br>

#### Minimal filtering to reduce the size of the dataset
```{r}
keep <- rowSums(counts(dds) >= 5) >= 4
table(keep)
dds <- dds[keep,]
```
We do not need to retain genes if they do not have a count of 5 or more for 4 or more samples as these genes will have no statistical power to detect differences, and no information to compute distances between samples.

<br>

#### Basic Exploratory Analysis
```{r}
boxplot(log10(counts(dds)+1))
```

<br>

#### Compute size factors to normalize counts for variation in sequencing depth among samples
The main function of DESeq2 includes calculation of size factors, but we are calculating these manually, so that the normalized counts are available for plotting.
```{r}
dds <- estimateSizeFactors(dds)
boxplot(log10(counts(dds,normalized=TRUE)+1))
```
Taking the logarithm of counts plus a pseudocount of 1 is a common transformation, but it tends to inflate the sampling variance of low counts such that it is even larger than biological variation across groups of samples. DESeq2 therefore provides transformations which produce log-scale data such that the systematic trends have been removed. Recommended transformation is the variance-stabilizing transformation, or VST, and it can be called with the vst function.

<br>

#### Perform Variant Stabilizing Transformation (VST)
```{r}
vsd <- vst(dds)
class(vsd)
```
This function does not return a DESeqDataSet, because it does not return counts, but instead continuous values (on the log2 scale). We can access the transformed data with assay:

<br>

#### Access the VSD-tranformed counts
```{r}
assay(vsd)[1:3,1:3]
```

<br>

## Exploratory Analysis

#### Visualize the VSD-tranformed data using Principal components Analysis
```{r}
#plotPCA(vsd, "dex") # An option for a simple plot

# Prepare data for ggplot
pcaData <- plotPCA(vsd, intgroup = c( "dex", "cell"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

# Plot using ggplot
ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
```


<br>

## Differential Expression Analysis

#### Apply DESeq Function on un-normalized counts
```{r}
dds <- DESeq(dds)
res <- results(dds)
```

<br>

# See the top genes (order by p-value)
```{r}
head(res[order(res$pvalue),])
```
<br>

# See the top genes (order by p-value)
```{r}
plotMA(res, ylim=c(-5,5), colSig = "firebrick3")
```
For more informative visualization and more accurate ranking of genes by effect size, recommend to use DESeq2’s functionality for shrinking LFCs. For this, the apeglm shrinkage estimator is available in DESeq2’s lfcShrink function.

# See the top genes (order by p-value)
```{r}
res2 <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm")
```
<br>

#### MA plots with/without Shrinkage:
```{r}
par(mfrow=c(1,2))
plotMA(res, ylim=c(-3,3), main="No shrinkage", colSig = "firebrick3")
plotMA(res2, ylim=c(-3,3), main="apeglm", colSig = "firebrick3")
```
<br>

#### Specifying a minimum biologically meaningful effect size by choosing an LFC cutoff
```{r}
res.lfc <- results(dds, lfcThreshold=1)
res.lfc2 <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm",
                      lfcThreshold=1)
```

<br>

#### MA plots with/without Shrinkage, with LFC cutoffs added
```{r}
par(mfrow=c(1,2))
plotMA(res.lfc, ylim=c(-5,5), main="No shrinkage, LFC test", colSig = "firebrick3")
plotMA(res.lfc2, ylim=c(-5,5), main="apeglm, LFC test", alpha=0.01, colSig = "firebrick3")
```

<br>

## Annotations

#### Will use the AnnotationHub package to attach additional information to the results table
```{r}
ah <- AnnotationHub()
query(ah, c("OrgDb","Homo sapiens"))
```

<br>

#### Retrieve record for Homo Sapiens
```{r}
hs <- ah[["AH79577"]]
hs
```


<br>

### Mapping IDs

#### Rownames of the results table are Ensembl IDs, and most of these are entries in OrgDb (although thousands are not).
```{r}
columns(hs)
table(rownames(res) %in% keys(hs, "ENSEMBL"))
```
We can use the mapIds function to add gene symbols, using ENSEMBL as the keytype, and requesting the column SYMBOL.

<br>

#### 
```{r}

```

<br>

#### 
```{r}
res$symbol <- mapIds(hs, rownames(res), column="SYMBOL", keytype="ENSEMBL")
head(res)
```
<br>

### Building Reports
```{r}
tmp <- tempdir() # you would instead use a meaningful path here
rep <- HTMLReport(shortName="airway", title="Airway DGE",
                  basePath=tmp, reportDirectory="report")
publish(res, rep, dds, n=20, make.plots=TRUE, factor=dds$dex)
finish(rep)
```
```{r}
browseURL(file.path(tmp,"report","airway.html"))

```

